contracts/Registry.sol

```diff
@@ -0,0 +1,121 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.24;
+
+import {AccessControlDefaultAdminRules} from '@openzeppelin/contracts/access/extensions/AccessControlDefaultAdminRules.sol';
+import {SafeERC20} from '@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol';
+import {IERC20} from '@openzeppelin/contracts/token/ERC20/IERC20.sol';
+
+contract Registry is AccessControlDefaultAdminRules {
+  using SafeERC20 for IERC20;
+
+  event ReferrerRegistered(
+    string indexed referrerId,
+    string[] protocolIds,
+    uint256[] rewardRates,
+    address rewardAddress
+  );
+
+  event ReferralRegistered(
+    string indexed protocolId,
+    string indexed referrerId,
+    address indexed userAddress
+  );
+
+  error ReferrerNotRegistered(string protocolId, string referrerId);
+
+  error UserAlreadyRegistered(
+    string protocolId,
+    string referrerId,
+    address userAddress
+  );
+
+  struct User {
+    uint256 timestamp;
+  }
+
+  struct Referrer {
+    address rewardAddress;
+    uint256 rewardRate;
+    address[] userAddresses;
+  }
+
+  struct Protocol {
+    string[] referrers;
+  }
+
+  mapping(string => Protocol) private _protocols;
+  mapping(string => mapping(string => Referrer)) private _referrers;
+  mapping(string => mapping(address => User)) private _usersByProtocol;
+
+  constructor(
+    address owner,
+    uint48 transferDelay
+  ) AccessControlDefaultAdminRules(transferDelay, owner) {}
+
+  function registerReferrer(
+    string calldata _referrerId,
+    string[] calldata _protocolIds,
+    uint256[] calldata _rewardRates,
+    address _rewardAddress
+  ) external onlyRole(DEFAULT_ADMIN_ROLE) {
+    for (uint256 i = 0; i < _protocolIds.length; i++) {
+      _referrers[_protocolIds[i]][_referrerId] = Referrer(
```

Comment:
A `Referrer` struct is not actually a "global" object per referrer; it seems to me to be implicitly tied to a particular protocol by virtue of its placement in the `_referrers` mapping. However, it does contain a piece of state, `rewardAddress`, which _is_ global/unique to a particular referrer. With the current state setup, there is no real source of truth for `rewardAddress` for a particular referrer, since it may be duplicated several times among different `Referrer` objects. Kind of a nitpick, but I might suggest moving `rewardAddress` to a mapping from referrer ID to reward address.

Also (as I mention in a previous comment), we currently have no way to retrieve reward address from the contract state for a particular referrer.